lib/kubernetes/tools.rb

```diff
@@ -0,0 +1,42 @@
+# frozen_string_literal: true
+
+module Kubernetes
+  module Tools
+
+    module_function
+
+    def running_on_kubernetes?
+      Dir.exist?('/run/secrets/kubernetes.io') || ENV.key?('KUBERNETES_PORT')
+    end
+
+    def cpu_shares
+      # This check is from https://github.com/kubernetes/kubernetes/blob/release-1.27/test/e2e/node/pod_resize.go#L305-L314
+      # alternatively, this method can be used: https://kubernetes.io/docs/concepts/architecture/cgroups/#check-cgroup-version
+      # (`stat -fc %T /sys/fs/cgroup/` returns `cgroup2fs` or `tmpfs`)
+      if File.exist?('/sys/fs/cgroup/cgroup.controllers')
+        # Cgroups v2
+        # Using the formula from https://github.com/kubernetes/kubernetes/blob/release-1.27/pkg/kubelet/cm/cgroup_manager_linux.go#L570-L574
+        weight = Integer(File.read('/sys/fs/cgroup/cpu.weight'))
+        shares = (((weight - 1) * 262142) / 9999) + 2
+      else
+        # Cgroups v1
+        shares = Integer(File.read('/sys/fs/cgroup/cpu/cpu.shares'))
+      end
+      shares
+    rescue StandardError => exception
+      warn "WARNING: Caught exception: #{exception}"
+    end
+
+    def available_cpus
+      (cpu_shares / 1024.0).ceil
+    rescue StandardError => exception
+      warn "WARNING: Caught exception: #{exception}"
+    end
+
+    def kubernetes_cpu_request
+      available_cpus if running_on_kubernetes?
```

Comment:
Maybe we shouldn't care whether we run on k8s. If we see cgroups cpu limit, then use that, otherwise do not. Even if we run directly under podman/docker or locally, it makes sense to use the `cgroups` value if such is present.

Reply:
looks great overall!

Reply:
Yeah, I guess it makes sense.
I'll refactor

Reply:
@akostadinov take a look at https://github.com/3scale/porta/pull/3565/commits/742b5201fa562430a081e3b83d99a93a9f8f2708 please

One potential drawback is that if the image is run outside of K8S, it might be that `/sys/fs/cgroup/cpu.weight` could be missing (even if cgroups v2 is used), because of how the cgroup directory is structured... So the following warning can be observed:
```
WARNING: Caught exception getting CPU shares: No such file or directory @ rb_sysopen - /sys/fs/cgroup/cpu.weight
```


But I think it's OK, I guess we'll rarely run the bare container outside of K8s, and it will be only for testing purposes.
When running locally we won't see this warning because a different unicorn config is used.